bindings/python/repository.py

```diff
@@ -0,0 +1,185 @@
+import json
+import os
+import tarfile
+import typing as t
+from abc import ABC, abstractmethod
+from functools import partial
+from urllib.parse import urlparse
+
+import requests
+from appdirs import AppDirs
+
+from .typing_utils import URL, PathLike
+from .utils import download_resource, patch_marian_for_bergamot
+
+APP = "bergamot"
+
+
+class Repository(ABC):
+    """
+    An interface for several repositories. Intended to enable interchangable
+    use of translateLocally and Mozilla repositories for usage through python.
+    """
+
+    @property
+    @abstractmethod
+    def name(self):
+        pass
+
+    @abstractmethod
+    def update(self):
+        """Updates the model list"""
+        pass
+
+    @abstractmethod
+    def models(self) -> t.List[str]:
+        """returns identifiers for available models"""
+        pass
+
+    @abstractmethod
+    def model(self, model_identifier: str) -> t.Any:
+        """returns entry for the  for available models"""
+        pass
+
+    @abstractmethod
+    def modelConfigPath(self, model_identifier: str) -> str:
+        """returns modelConfigPath for for a given model-identifier"""
+        pass
+
+    @abstractmethod
+    def download(self, model_identifier: str):
+        pass
+
+
+class TranslateLocallyLike(Repository):
+    """
+    This class implements Repository to fetch models from translateLocally.
+    AppDirs is used to standardize directories and further specialization
+    happens with translateLocally identifier.
+    """
+
+    def __init__(self, name, url):
+        self.url = url
+        self._name = name
+        appDir = AppDirs(APP)
+        f = lambda *args: os.path.join(*args, self._name)
+        self.dirs = {
+            "cache": f(appDir.user_cache_dir),
+            "config": f(appDir.user_config_dir),
+            "data": f(appDir.user_data_dir),
+            "archive": f(appDir.user_data_dir, "archives"),
+            "models": f(appDir.user_data_dir, "models"),
+        }
+
+        for directory in self.dirs.values():
+            os.makedirs(directory, exist_ok=True)
+
+        self.models_file_path = os.path.join(self.dirs["config"], "models.json")
+        self.update()
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    def update(self) -> None:
+        inventory = requests.get(self.url).text
+        with open(self.models_file_path, "w+") as models_file:
+            models_file.write(inventory)
+        self.data = json.loads(inventory)
```

Comment:
I have updated the models.json and model_info.json to contain repository information. Eg: https://translatelocally.com/models.json
https://github.com/browsermt/students/blob/master/enfa/enfa.student.tiny11/model_info.json#L7